### PR TITLE
feat(metadata): add ETH Swarm (bzz://) support for token metadata

### PR DESCRIPTION
**Pull request by entin-hun**
## Summary

Adds first-class support for **ETH Swarm** hashes in the token metadata pipeline, mirroring the existing Arweave integration.

The implementation works with **any content stored on Swarm**, not just `/bzz` uploads — because the entry point is the 64-character keccak-256 *hash* (content address), which is shared by:
- `/bzz/<hash>` — regular chunk uploads
- SOC (Single Owner Chunks) references
- Feed manifests (which resolve to a content-address under the hood)

A 200 response from the Swarm gateway is therefore sufficient proof that the content exists, regardless of how it was originally uploaded.

---

### `apps/explorer/lib/explorer/token/metadata_retriever.ex`

| Addition | Purpose |
|---|---|
| `@invalid_swarm_path` | Error constant (mirrors `@invalid_ipfs_path`) |
| `swarm_link/1` | Builds `<gateway>/bzz/<hash>/`; defaults to `https://gateway.ethswarm.org`; overridable via `config :indexer, :swarm, gateway_url:` |
| `swarm_headers/0` | Returns `Authorization: Bearer <token>` when `config :indexer, :swarm, bearer_token:` is set — supports private / paid gateway nodes |
| `valid_swarm_hash?/1` | Validates 64-char lowercase hex Swarm hash |
| `bzz://` clause in `fetch_from_ipfs_or_ar?/5` | Routes `bzz://<hash>[/path]` URIs |
| `/bzz/<hash>` clause in `fetch_from_ipfs_or_ar?/5` | Routes plain HTTPS Swarm gateway URLs |
| `fetch_from_swarm/2` + `fetch_from_swarm_if_valid_hash/2` | Fetcher (mirrors `fetch_from_arweave/2`) |
| `swarm?/1` predicate | Threaded through host-filtering bypass, header selection, and `process_result/3` |

### `apps/explorer/test/explorer/token/metadata_retriever_test.exs`

New describe blocks:
- `swarm_link/1` — default gateway, custom gateway, trailing-slash stripping
- `swarm_headers/0` — no token, bearer token configured
- `valid_swarm_hash?/1` — valid, embedded path, uppercase, too short/long, empty, non-binary
- ETH Swarm URI routing — `bzz://` scheme, plain HTTPS `/bzz/`, invalid hash error

---

## Configuration (optional)

```elixir
# config/runtime.exs  (or config.exs)
config :indexer, :swarm,
  gateway_url: System.get_env("INDEXER_SWARM_GATEWAY_URL", "https://gateway.ethswarm.org"),
  bearer_token: System.get_env("INDEXER_SWARM_GATEWAY_BEARER_TOKEN")
```

Both keys are optional — the module works with zero configuration using the public Swarm gateway.

---

## Related

Closes #<issue> (if any)

Follows the same pattern as the Arweave integration introduced in [this commit](https://github.com/blockscout/blockscout/blob/5b62fd2e34e8f9cbad6bdcaa8c41f28e187516b9/apps/explorer/lib/explorer/token/metadata_retriever.ex#L794).

## Summary by CodeRabbit

* **New Features**
  * Added ETH Swarm support for token metadata retrieval: recognizes bzz:// and gateway-style paths, normalizes and validates Swarm identifiers, routes fetches through a configurable gateway, and returns Swarm metadata consistently.
  * Optional gateway URL configuration and optional Bearer token authorization for gateway requests.

* **Tests**
  * Added coverage for gateway URL construction, header/authorization behavior, identifier validation, fetching (including deep paths), and error handling.

### CODE REVIEW

File: apps/explorer/lib/explorer/token/metadata_retriever.ex [hunk 1]
```diff
@@ -741,6 +797,22 @@ defmodule Explorer.Token.MetadataRetriever do
...
+            is_binary(host) -> host
+            true -> nil
+          end
+
+        fetch_from_swarm_if_valid_hash(uid, hex_token_id)
+
+      %URI{scheme: _, path: "/bzz/" <> resource_id} ->
+        # https://gateway.ethswarm.org/bzz/<hash>[/path] — strip trailing path
+        hash = resource_id |> String.split("/") |> List.first()
+        fetch_from_swarm_if_valid_hash(hash, hex_token_id)
```
[the remark this answers is not in the document]
Comment (entin-hun) [reply apps/explorer/lib/explorer/token/metadata_retriever.ex]: Good catch. Updated this route to preserve deep paths for HTTPS Swarm URLs (now forwarding the full `<hash>/...` resource_id) so metadata files under nested paths are fetched correctly.

File: apps/explorer/lib/explorer/token/metadata_retriever.ex [hunk 2]
```diff
@@ -798,6 +870,19 @@ defmodule Explorer.Token.MetadataRetriever do
     fetch_metadata_inner(arweave_url, ipfs_params, nil, hex_token_id)
   end
 
+  defp fetch_from_swarm(uid, hex_token_id) do
+    url = swarm_link(uid)
+    fetch_metadata_inner(url, [ipfs?: false, swarm?: true], nil, hex_token_id)
```
[the remark this answers is not in the document]
Comment (entin-hun) [reply apps/explorer/lib/explorer/token/metadata_retriever.ex:932]: Thanks. This is already handled in the PR code path: host-filter bypass now includes `!swarm?(ipfs_params)` in `fetch_metadata_from_uri/3`, so Swarm requests are treated consistently with IPFS/Arweave. I kept `ipfs?: false` for Swarm to preserve IPFS-specific header/query behavior separation.

File: apps/explorer/test/explorer/token/metadata_retriever_test.exs [hunk 3]
```diff
@@ -1275,4 +1275,147 @@ defmodule Explorer.Token.MetadataRetrieverTest do
...
+  end
+
+  describe "ETH Swarm URI routing" do
+    @valid_hash String.duplicate("b2", 32)
+    @swarm_metadata %{"name" => "Swarm NFT", "description" => "stored on Swarm"}
+
+    test "fetch_json resolves bzz:// URI scheme via Swarm gateway" do
+      Application.put_env(:indexer, :swarm, gateway_url: "https://gateway.ethswarm.org")
+
+      expected_url = "https://gateway.ethswarm.org/bzz/#{@valid_hash}/"
```
[the remark this answers is not in the document]
Comment (entin-hun) [reply apps/explorer/test/explorer/token/metadata_retriever_test.exs:1399]: Fixed. Added `on_exit/1` restoration for Swarm env in this test to prevent config leakage even when assertions fail.

File: apps/explorer/test/explorer/token/metadata_retriever_test.exs [hunk 4]
```diff
@@ -1275,4 +1275,147 @@ defmodule Explorer.Token.MetadataRetrieverTest do
...
+      assert {:ok, %{metadata: metadata}} =
+               MetadataRetriever.fetch_json({:ok, ["bzz://#{@valid_hash}"]})
+
+      assert metadata["name"] == "Swarm NFT"
+    end
+
+    test "fetch_json resolves https://gateway.ethswarm.org/bzz/<hash>/ URL" do
+      Application.put_env(:indexer, :swarm, gateway_url: "https://gateway.ethswarm.org")
+
+      url = "https://gateway.ethswarm.org/bzz/#{@valid_hash}/"
```
[the remark this answers is not in the document]
Comment (entin-hun) [reply apps/explorer/test/explorer/token/metadata_retriever_test.exs:1422]: Fixed here as well. Added `on_exit/1` restoration for the `:swarm` env in this test case too.

File: apps/explorer/lib/explorer/token/metadata_retriever.ex [hunk 5]
```diff
@@ -553,6 +554,36 @@ defmodule Explorer.Token.MetadataRetriever do
...
+  """
+  @spec swarm_link(uid :: any()) :: String.t()
+  def swarm_link(uid) do
+    base_url =
+      :indexer
+      |> Application.get_env(:swarm, [])
+      |> Keyword.get(:gateway_url, "https://gateway.ethswarm.org")
+      |> String.trim_trailing("/")
+
+    "#{base_url}/bzz/#{uid}/"
```
[the remark this answers is not in the document]
Comment (entin-hun) [reply apps/explorer/lib/explorer/token/metadata_retriever.ex]: Implemented. `swarm_link/1` now preserves deep paths without forcing a trailing slash, while still keeping the trailing slash for plain hash-only UIDs. Also added coverage for deep-path behavior in tests.

**Comment by entin-hun** [comment]
Addressed the review suggestions in commit f344a3bf8:

- Preserved deep paths for HTTPS Swarm gateway URLs () instead of truncating at hash.
- Updated  to avoid forcing a trailing slash when UID includes a path segment.
- Removed unnecessary  stub from Swarm routing test.
- Hardened Swarm env handling in tests with  restoration to avoid leakage.
- Replaced hardcoded User-Agent assertions by deriving defaults from .
- Added regression test covering deep-path Swarm URL fetch.

Note: I could not run  in this environment because Elixir tooling () is not installed here.

**Comment by entin-hun** [comment]
Follow-up status update (previous summary comment had shell-escaping glitches):

Addressed review suggestions in commit f344a3bf8.

- Preserved deep paths for HTTPS Swarm gateway URLs in the /bzz route instead of truncating at the hash.
- Updated swarm_link/1 to avoid forcing a trailing slash when UID already includes a path segment.
- Removed the unnecessary EthereumJSONRPC.Mox stub from the first Swarm routing test.
- Added on_exit restoration for :swarm env in tests to prevent config leakage.
- Replaced hardcoded User-Agent assertions by deriving defaults from MetadataRetriever.ar_headers/0.
- Added a regression test for deep-path Swarm URL fetching.

I could not run mix test in this environment because mix is not installed.

File: apps/explorer/lib/explorer/token/metadata_retriever.ex [hunk 6]
```diff
@@ -798,6 +875,19 @@ defmodule Explorer.Token.MetadataRetriever do
     fetch_metadata_inner(arweave_url, ipfs_params, nil, hex_token_id)
   end
 
+  defp fetch_from_swarm(uid, hex_token_id) do
+    url = swarm_link(uid)
+    fetch_metadata_inner(url, [ipfs?: false, swarm?: true], nil, hex_token_id)
```
[the remark this answers is not in the document]
Comment (entin-hun) [reply apps/explorer/lib/explorer/token/metadata_retriever.ex:932]: Implemented in d08cff909. Added !swarm?(ipfs_params) to the host-filter bypass in fetch_metadata_from_uri/3, so Swarm fetches now skip MetadataURIValidator.validate_uri/1 just like IPFS/Arweave and can reach configured gateway/bearer-token paths.

File: apps/explorer/lib/explorer/token/metadata_retriever.ex [hunk 7]
```diff
@@ -622,6 +659,31 @@ defmodule Explorer.Token.MetadataRetriever do
...
+  """
+  @spec swarm_headers() :: [{binary(), binary()}]
+  def swarm_headers do
+    swarm_params = Application.get_env(:indexer, :swarm, [])
+    bearer_token = Keyword.get(swarm_params, :bearer_token)
+
+    if bearer_token do
+      [{"Authorization", "Bearer #{bearer_token}"} | @default_headers]
+    else
+      @default_headers
```
[the remark this answers is not in the document]
Comment (entin-hun) [reply apps/explorer/lib/explorer/token/metadata_retriever.ex]: Implemented in 7068effd3. Updated `swarm_headers/0` to only prepend Authorization when `bearer_token` is a non-empty, non-whitespace binary (`is_binary(bearer_token) and String.trim(bearer_token) != ""`). Also added tests for empty and whitespace tokens to ensure we return default headers in those cases.

**Comment by vbaranov** [comment]
Thank you for contribution! I finalized changes to fix CI in https://github.com/blockscout/blockscout/pull/14446. It will be included into next major release.

**Comment by entin-hun** [comment]
> Thank you for contribution! I finalized changes to fix CI in #14446. It will be included into next major release.

Thanks for being open!

**Comment by parkintr** [comment]
> ## Summary
> 
> Adds first-class support for **ETH Swarm** hashes in the token metadata pipeline, mirroring the existing Arweave integration.
> 
> The implementation works with **any content stored on Swarm**, not just `/bzz` uploads — because the entry point is the 64-character keccak-256 *hash* (content address), which is shared by:
> - `/bzz/<hash>` — regular chunk uploads
> - SOC (Single Owner Chunks) references
> - Feed manifests (which resolve to a content-address under the hood)
> 
> A 200 response from the Swarm gateway is therefore sufficient proof that the content exists, regardless of how it was originally uploaded.
> 
> ---
> 
> ## Changes
> 
> ### `apps/explorer/lib/explorer/token/metadata_retriever.ex`
> 
> | Addition | Purpose |
> |---|---|
> | `@invalid_swarm_path` | Error constant (mirrors `@invalid_ipfs_path`) |
> | `swarm_link/1` | Builds `<gateway>/bzz/<hash>/`; defaults to `https://gateway.ethswarm.org`; overridable via `config :indexer, :swarm, gateway_url:` |
> | `swarm_headers/0` | Returns `Authorization: Bearer <token>` when `config :indexer, :swarm, bearer_token:` is set — supports private / paid gateway nodes |
> | `valid_swarm_hash?/1` | Validates 64-char lowercase hex Swarm hash |
> | `bzz://` clause in `fetch_from_ipfs_or_ar?/5` | Routes `bzz://<hash>[/path]` URIs |
> | `/bzz/<hash>` clause in `fetch_from_ipfs_or_ar?/5` | Routes plain HTTPS Swarm gateway URLs |
> | `fetch_from_swarm/2` + `fetch_from_swarm_if_valid_hash/2` | Fetcher (mirrors `fetch_from_arweave/2`) |
> | `swarm?/1` predicate | Threaded through host-filtering bypass, header selection, and `process_result/3` |
> 
> ### `apps/explorer/test/explorer/token/metadata_retriever_test.exs`
> 
> New describe blocks:
> - `swarm_link/1` — default gateway, custom gateway, trailing-slash stripping
> - `swarm_headers/0` — no token, bearer token configured
> - `valid_swarm_hash?/1` — valid, embedded path, uppercase, too short/long, empty, non-binary
> - ETH Swarm URI routing — `bzz://` scheme, plain HTTPS `/bzz/`, invalid hash error
> 
> ---
> 
> ## Configuration (optional)
> 
> ```elixir
> # config/runtime.exs  (or config.exs)
> config :indexer, :swarm,
>   gateway_url: System.get_env("INDEXER_SWARM_GATEWAY_URL", "https://gateway.ethswarm.org"),
>   bearer_token: System.get_env("INDEXER_SWARM_GATEWAY_BEARER_TOKEN")
> ```
> 
> Both keys are optional — the module works with zero configuration using the public Swarm gateway.
>
> parkintr88@gmail.com
> ---
>
> +66646387039
> ## Related
>
> PARKIN  CUMLIE.
> Closes #<issue> (if any)
> PARKIN CUMLIIEM
> Follows the same pattern as the Arweave integration introduced in [this commit](https://github.com/blockscout/blockscout/blob/5b62fd2e34e8f9cbad6bdcaa8c41f28e187516b9/apps/explorer/lib/explorer/token/metadata_retriever.ex#L794).
> PARKIN CUMLIEM
> 
> ## Summary by CodeRabbit
> Parkintr88
> * **New Features**
>   * Added ETH Swarm support for token metadata retrieval: recognizes bzz:// and gateway-style paths, normalizes and validates Swarm identifiers, routes fetches through a configurable gateway, and returns Swarm metadata consistently.
>   * Optional gateway URL configuration and optional Bearer token authorization for gateway requests.
> 
> * **Tests**
>   * Added coverage for gateway URL construction, header/authorization behavior, identifier validation, fetching (including deep paths), and error handling.
> /withdraw Amount Token WithdrawAddress Memo (optional)httmytonwallet.io/
